dmutils/views.py

```diff
@@ -0,0 +1,223 @@
+from abc import ABCMeta, abstractmethod
+import csv
+import enum
+from flask import abort, request, Response
+from flask.views import View
+from io import BytesIO
+from odf.style import TextProperties, TableRowProperties, TableColumnProperties, TableCellProperties, FontFace
+import six
+
+from dmutils import csv_generator
+from dmutils import ods
+
+
+@six.add_metaclass(ABCMeta)
+class DownloadFileView(View):
+    """An abstract base class appropriate for subclassing in the frontend apps when the user needs to be able to
+    download some data as a CSV or ODS file. All abstract methods must be implemented on the subclass; all other
+    methods should be able to remain as-is to support handling and dispatching the request."""
+
+    FILETYPES = enum.Enum('Filetypes', ['CSV', 'ODS'])
+
+    def __init__(self, **kwargs):
+        self.request = request
+
+        self.data_api_client = None
+        self.search_api_client = None
+        self.content_loader = None
+
+        super(View, self).__init__(**kwargs)
+
+    @abstractmethod
+    def _init_hook(self, **kwargs):
+        """Hook into the start of the request to associate required clients/loaders with the instance and perform any
+        other logic required for the start of the request."""
+        self.data_api_client = None
+        self.search_api_client = None
+        self.content_loader = None
+
+    def _post_request_hook(self, response=None, file_context=None, **kwargs):
+        """A hook that can be used to implement any logic that should be run after the main download file process,
+        and immediately before sending the response to the user."""
+        pass
+
+    @abstractmethod
+    def get_file_context(self, **kwargs):
+        """Must return a dictionary containing any data required by the generation routines.
+
+        Required keys:
+         * `filename`: without a file extension"""
+        return {'filename': 'my-download'}
+
+    @abstractmethod
+    def determine_filetype(self, file_context=None, **kwargs):
+        """Logic to tell the View which filetype to generate; must return a choice from DownloadFileView.FILETYPES"""
+        return DownloadFileView.FILETYPES.ODS
+
+    @abstractmethod
+    def generate_csv_rows(self, file_context):
+        """Must return a nested list of rows+cells, [[heading1, heading2], [data1, data2], ...]"""
+        return [['Heading 1', 'Heading 2'], ['Row 1, Column 1', 'Row 1, Column 2']]
```

Comment:
Do they *have to* be lists or just sequences?

Reply:
You're right - just sequences.